File: backend/app/pipeline/notify.py
"""Web Push notifications (pywebpush + VAPID), sent to the browser
subscription stored on profiles.push_subscription (jsonb: the standard
PushSubscription JSON: {endpoint, keys:{p256dh,auth}}).
"""

import json
import logging

from pywebpush import WebPushException, webpush

from app.config import settings

logger = logging.getLogger("notify")


async def send_push(push_subscription: dict | str | None, title: str, body: str, data: dict | None = None) -> bool:
    """Returns False (and the caller should drop the stored subscription) if
    the push service reports the subscription is gone (404/410)."""
    if not push_subscription:
        return True
    if isinstance(push_subscription, str):
        push_subscription = json.loads(push_subscription)

    payload = json.dumps({"title": title, "body": body, "data": data or {}})
    try:
        webpush(
            subscription_info=push_subscription,
            data=payload,
            vapid_private_key=settings.vapid_private_key,
            vapid_claims={"sub": settings.vapid_subject},
        )
        return True
    except WebPushException as e:
        status = getattr(e.response, "status_code", None)
        if status in (404, 410):
            return False
        logger.warning("push failed: %s", e)
        return True

File: web/src/server/pipeline/notify.ts
import "server-only";
import webpush from "web-push";
import { config } from "../config";
import { sql } from "../db";

/** Web Push (web-push npm + VAPID) to the subscription stored on
 * profiles.push_subscription (jsonb: standard PushSubscription JSON).
 * Port of backend/app/pipeline/notify.py. */

export interface PushSubscriptionLike {
  endpoint: string;
  keys: { p256dh: string; auth: string };
}

/** Sends one push message. Returns false if the push service reports the
 * subscription is gone (404/410) — caller should drop the stored
 * subscription in that case. */
export async function sendPush(
  subscription: PushSubscriptionLike | null | undefined,
  title: string,
  body: string,
  data: Record<string, unknown> = {},
): Promise<boolean> {
  if (!subscription) return true;
  webpush.setVapidDetails(config.vapidSubject, config.vapidPublicKey, config.vapidPrivateKey);
  try {
    await webpush.sendNotification(
      subscription as webpush.PushSubscription,
      JSON.stringify({ title, body, data }),
    );
    return true;
  } catch (e) {
    const status = (e as { statusCode?: number }).statusCode;
    if (status === 404 || status === 410) return false;
    console.warn("push failed", e);
    return true;
  }
}

/** Loads the user's push subscription and sends; drops the stored
 * subscription if the push service reports it's gone. */
export async function notifyUser(
  userId: string,
  title: string,
  body: string,
  data: Record<string, unknown> = {},
): Promise<void> {
  const db = sql();
  const rows = await db<{ push_subscription: PushSubscriptionLike | null }[]>`
    select push_subscription from profiles where user_id=${userId}
  `;
  const sub = rows[0]?.push_subscription;
  if (!sub) return;
  const ok = await sendPush(sub, title, body, data);
  if (!ok) {
    await db`update profiles set push_subscription=null where user_id=${userId}`;
  }
}
